In [11]:
products = [
    {"name": "키보드", "price":45000, "stock":12, "category":"전자기기"},
    {"name": "마우스", "price":45000, "stock":0, "category":"전자기기"},
    {"name": "모니터", "price":45000, "stock":4, "category":"전자기기"},
    {"name": "책상", "price":45000, "stock":0, "category":"가구"},
    {"name": "의자", "price":45000, "stock":7, "category":"가구"},
    {"name": "키보드", "price":45000, "stock":5, "category":"전자기기"},
    {"name": "마우스", "price":45000, "stock":2, "category":"전자기기"},
    {"name": "의자", "price":45000, "stock":3, "category":"가구"},
]

print(products)

[{'name': '키보드', 'price': 45000, 'stock': 12, 'category': '전자기기'}, {'name': '마우스', 'price': 45000, 'stock': 0, 'category': '전자기기'}, {'name': '모니터', 'price': 45000, 'stock': 4, 'category': '전자기기'}, {'name': '책상', 'price': 45000, 'stock': 0, 'category': '가구'}, {'name': '의자', 'price': 45000, 'stock': 7, 'category': '가구'}, {'name': '키보드', 'price': 45000, 'stock': 5, 'category': '전자기기'}, {'name': '마우스', 'price': 45000, 'stock': 2, 'category': '전자기기'}, {'name': '의자', 'price': 45000, 'stock': 3, 'category': '가구'}]


In [13]:
for_sales = []

found = False

for product in products:
    if product["stock"] >= 1:
        found = False
        for sale in for_sales:
            if sale["name"] == product["name"]:
                sale["stock"] += product["stock"]
                found = True
                break
        if not found:
            for_sales.append(product.copy())
            
for_sales    

[{'name': '키보드', 'price': 45000, 'stock': 17, 'category': '전자기기'},
 {'name': '모니터', 'price': 45000, 'stock': 4, 'category': '전자기기'},
 {'name': '의자', 'price': 45000, 'stock': 10, 'category': '가구'},
 {'name': '마우스', 'price': 45000, 'stock': 2, 'category': '전자기기'}]

### 실습 1. Excel 파일 불러오기

해야 할 일

pandas를 이용해 Excel 파일을 DataFrame으로 불러옵니다.

먼저 코드부터 작성하기보다 인공지능에게 현재 상황과 원하는 작업을 설명해 봅니다.


In [8]:
import pandas as pd

# 파일이 있는 전체 경로(폴더명 포함)를 적어줍니다.
df = pd.read_excel('notebooks/book-text-ml/books.xlsx')

# 앞의 5개 행 확인
df.head()

c:\dev\llm-data-analysis-course\.venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


### 📚 도서 베스트셀러 데이터 분석 요약

* **불러온 파일명**: `books.xlsx` (교보문고 베스트셀러 상품 리스트)
* **데이터 형태**: pandas DataFrame (`df`)
* **확인 내용**: 데이터의 상위 5개 행(`df.head()`) 조회 완료

| 순번 | 주요 특징 / 의미 |
| :--- | :--- |
| **1** | pandas 라이브러리를 이용해 엑셀 데이터를 정상적으로 메모리에 적재함 |
| **2** | `df.head()`를 통해 전체 데이터의 구조, 컬럼명(책 제목, 저자 등), 상위 데이터의 형태를 빠르게 파악함 |

### 실습 2. 데이터의 크기와 컬럼 확인하기

Excel 파일을 불러왔으면 바로 전처리를 시작하지 않습니다.
 
먼저 어떤 데이터인지 확인합니다.

확인할 내용

다음 내용을 확인해 봅니다.

In [14]:
# 1. 전체 행과 컬럼 개수 확인 (결과: (행 개수, 열 개수))
print("--- 데이터 크기 ---")
print(df.shape)

# 2. 컬럼 이름 목록 확인
print("\n--- 컬럼 이름 목록 ---")
print(df.columns)

# 3. 각 컬럼의 자료형 확인 (int: 정수, float: 실수, object: 문자열/복합데이터 등)
print("\n--- 각 컬럼의 자료형 ---")
print(df.dtypes)

# 4. 각 컬럼의 결측치(비어있는 값) 개수 확인
# isna()로 결측치 여부를 확인하고, sum()으로 그 개수를 모두 더해줍니다.
print("\n--- 결측치 개수 ---")
print(df.isna().sum())

# 5. 앞의 5행 확인
# 주피터 노트북에서는 맨 마지막 줄에 작성하면 깔끔한 표 형태로 출력됩니다.
df.head()

--- 데이터 크기 ---
(199, 13)

--- 컬럼 이름 목록 ---
Index(['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트',
       '인물', '출판사', '발행(출시)일자', '분야'],
      dtype='str')

--- 각 컬럼의 자료형 ---
순위          int64
상품코드        int64
판매상품 ID       str
상품명           str
정가            str
판매가           str
할인율           str
적립율           str
적립예정포인트       str
인물            str
출판사           str
발행(출시)일자    int64
분야            str
dtype: object

--- 결측치 개수 ---
순위          0
상품코드        0
판매상품 ID     0
상품명         0
정가          0
판매가         0
할인율         0
적립율         0
적립예정포인트     0
인물          0
출판사         0
발행(출시)일자    0
분야          0
dtype: int64


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


In [13]:
df.isna().sum().sum() # sum을 두 번 하는 이유 : 한 번 하면 컬럼별로 나옴.

np.int64(0)

### 📊 데이터 기본 구조 확인 요약

* **데이터 크기**: 총 199개의 도서 데이터(행)와 13개의 정보(열)로 구성되어 있습니다.
* **결측치(빈 값)**: 비어있는 데이터가 하나도 없는 아주 깨끗한 상태입니다. (전체 결측치 0개)
* **💡 분석 시 참고사항 (자료형)**:
  * `정가`, `판매가`, `할인율` 등 숫자여야 할 데이터들이 현재 **문자열(str)**로 되어 있습니다. (아마도 쉼표나 기호가 포함되어 있을 가능성이 높습니다.)
  * 추후 가격을 계산하거나 비교 분석하려면, 이 컬럼들을 숫자형(int 또는 float)으로 변경해 주는 전처리 작업이 필요합니다.

### 실습 3. 분석에 필요한 컬럼만 선택하기

원본 데이터의 모든 컬럼이 이후 분석에 필요한 것은 아닙니다.

이번 과정에서는 다음 컬럼을 먼저 선택합니다.

In [15]:
# 1. 원하는 컬럼만 선택해서 새로운 데이터프레임(df_books) 만들기
df_books = df[['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야']]

# 2. 데이터가 잘 추려졌는지 앞의 5개 행을 출력해서 확인하기
df_books.head()

,순위,판매상품 ID,상품명,판매가,인물,출판사,발행(출시)일자,분야
0,1,S000000610612,소년이 온다,"13,500",한강,창비,20140519,소설
1,2,S000001632467,모순,"11,700",양귀자,쓰다,20130401,소설
2,3,S000216262757,결국 국민이 합니다,"19,800",이재명,오마이북,20250415,정치/사회
3,4,S000216111714,혼모노,"16,200",성해나,창비,20250328,소설
4,5,S000200550189,급류,"12,600",정대건,민음사,20221222,소설


### 📝 데이터 컬럼 선택 결과 요약

* **추출 완료**: 전체 데이터 중 분석에 필요한 **8개의 핵심 컬럼만 성공적으로 추려내어** 새로운 데이터프레임(`df_books`)을 만들었습니다.
* **상위 5개 데이터 확인 (`head()`)**:
  * 1위 '소년이 온다(한강)', 2위 '모순(양귀자)' 등 도서 정보가 정상적으로 매칭되어 있습니다.
  * 상위권 5권 중 4권이 '소설' 분야입니다.
* **💡 분석 시 주의할 점**:
  * `판매가` 데이터를 보면 `13,500`처럼 숫자 사이에 **쉼표(,)**가 포함되어 있습니다. 
  * 컴퓨터는 이를 계산 가능한 숫자가 아닌 '문자열'로 인식하므로, 나중에 평균 가격 등을 계산하려면 쉼표를 없애고 숫자형으로 변환하는 전처리 작업이 필요할 수 있습니다.

### 실습 4. 컬럼 이름을 사용하기 쉽게 정리하기

원본 컬럼 이름은 사람이 읽기에는 좋지만 코드에서 반복해서 사용하기에는 조금 길 수 있습니다.


이번 실습에서는 다음처럼 간단하게 정리할 수 있습니다.

In [16]:
# 1. rename()을 사용하여 컬럼 이름 변경하기
# 딕셔너리 형태 { '기존 이름': '새 이름' }로 짝을 지어 입력합니다.
df_books = df_books.rename(columns={
    '판매상품 ID': '판매상품ID',
    '인물': '저자',
    '발행(출시)일자': '발행일'
})

# 2. 변경된 컬럼 이름 확인하기
print(df_books.columns)

Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')


### 🏷️ 컬럼 이름 변경 결과 요약

* **변경 완료**: 3개의 컬럼 이름이 의도한 대로 훨씬 직관적이고 간결하게 변경되었습니다.
  * `판매상품 ID` ➡️ **`판매상품ID`** (공백 제거로 코딩 시 오류 방지)
  * `인물` ➡️ **`저자`** (데이터의 의미가 더 명확해짐)
  * `발행(출시)일자` ➡️ **`발행일`** (이름이 짧아져서 다루기 편해짐)
* **현재 컬럼 목록**: 총 8개 (`순위`, `판매상품ID`, `상품명`, `판매가`, `저자`, `출판사`, `발행일`, `분야`)
* **의미**: 헷갈리기 쉽거나 긴 이름들을 깔끔하게 정리하여, 앞으로 데이터를 분석하고 코드를 작성하기가 훨씬 수월해졌습니다.

### 실습 5. 결측치 확인하기

결측치는 값이 비어 있는 데이터를 의미합니다.

데이터 분석에서는 결측치를 무조건 삭제하지 않습니다.

먼저 어디에 얼마나 있는지 확인하고, 분석 목적에 맞게 처리 방법을 결정합니다.

In [17]:
# 1. 전체 컬럼의 결측치(빈 데이터) 개수 확인하기
print("--- 전체 컬럼 결측치 개수 ---")
print(df_books.isna().sum())

# 2. 결측치가 있는 컬럼만 쏙 골라서 보기
print("\n--- 결측치가 있는 컬럼만 보기 ---")
missing_count = df_books.isna().sum() 
print(missing_count[missing_count > 0])

--- 전체 컬럼 결측치 개수 ---
순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

--- 결측치가 있는 컬럼만 보기 ---
Series([], dtype: int64)


### 🔍 결측치(빈 데이터) 확인 결과 요약

* **결측치 없음**: 전체 8개 컬럼 모두 누락된 데이터(결측치)가 **0개**로 확인되었습니다.
* **필터링 결과**: 결측치가 있는 컬럼만 뽑아본 결과 아무것도 나오지 않았습니다(`Series([])`).
* **의미**: 현재 데이터에 비어있는 칸이 하나도 없는 아주 깨끗한 상태입니다! 빈칸을 채우거나 지우는 복잡한 '결측치 처리' 작업 없이, 안심하고 바로 다음 분석 단계로 넘어가셔도 됩니다.

실습 6. 판매가를 숫자로 바꾸기


Excel에서 판매가가 다음과 같이 저장되어 있을 수 있습니다.

In [19]:
# 1. 판매가 컬럼의 쉼표(,) 제거하기
df_books['판매가'] = df_books['판매가'].str.replace(',', '')

# 2. 판매가 컬럼을 숫자형(정수)으로 변환하기
df_books['판매가'] = df_books['판매가'].astype(int)

# 3. 변환이 잘 되었는지 자료형(dtype) 확인하기
print(df_books['판매가'].dtype)

int64


### 💰 판매가 데이터 전처리 결과 요약

* **변환 완료**: '판매가' 컬럼의 자료형이 정상적으로 **정수형(`int64`)**으로 변경되었습니다.
* **처리 내용**:
  1. 가격 데이터에 포함되어 있던 쉼표(`,`) 기호를 모두 제거했습니다.
  2. 컴퓨터가 문자열로 인식하던 데이터를 계산 가능한 숫자(정수)로 바꾸었습니다.
* **의미**: 이제 '판매가' 데이터를 활용해 도서들의 평균 가격을 구하거나, 비싼 순서대로 정렬하는 등 본격적인 수학적 계산과 데이터 분석이 가능해졌습니다!

### 실습 7. 발행일을 날짜형으로 바꾸기

원본 데이터의 발행일은 다음처럼 보일 수 있습니다.

In [20]:
import pandas as pd

# 1. 발행일 컬럼을 날짜형(datetime)으로 변환하기
# format='%Y%m%d'를 통해 현재 데이터가 '연도(4자리)월(2자리)일(2자리)' 형태라고 알려줍니다.
df_books['발행일'] = pd.to_datetime(df_books['발행일'], format='%Y%m%d')

# 2. 변환이 잘 되었는지 자료형(dtype) 확인
print("--- 자료형 확인 ---")
print(df_books['발행일'].dtype)

# 3. 변환된 데이터 앞의 5개 확인
print("\n--- 데이터 확인 ---")
print(df_books['발행일'].head())

--- 자료형 확인 ---
datetime64[us]

--- 데이터 확인 ---
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]


### 📅 날짜(발행일) 데이터 변환 결과 요약

* **자료형 변경**: '발행일' 컬럼이 날짜와 시간을 다루는 전용 자료형인 **`datetime64`** 타입으로 완벽하게 변환되었습니다.
* **형식 변경**: `20140519`처럼 하나로 붙어있던 숫자들이 `2014-05-19`처럼 연-월-일 사이에 하이픈(-)이 들어간, 훨씬 읽기 편한 진짜 날짜 형태로 바뀌었습니다.
* **의미**: 이제 컴퓨터가 이 데이터를 '시간'으로 이해하기 때문에, "2024년에 출간된 책만 모아보기"나 "출간된 지 몇 년 지났는지 계산하기" 등 날짜를 활용한 강력한 분석이 가능해졌습니다!

### 실습 8. 문자열 앞뒤 공백 정리하기

상품명, 저자, 출판사, 분야는 문자열 데이터입니다.

문자열 앞뒤에 불필요한 공백이 있으면 같은 값도 서로 다른 데이터처럼 처리될 수 있습니다.

예를 들어 다음 두 값은 사람이 보기에는 같지만 문자열로는 다릅니다.

In [21]:
# 1. 공백을 제거할 문자열 컬럼 이름들을 리스트로 묶어줍니다.
text_columns = ['상품명', '저자', '출판사', '분야']

# 2. for 반복문을 이용해 컬럼을 하나씩 꺼내면서 공백을 제거합니다.
for col in text_columns:
    # str.strip(): 문자열(str)의 양 끝에 있는 공백을 잘라냅니다(strip).
    df_books[col] = df_books[col].str.strip()

# 3. 처리가 끝난 후, 해당 컬럼들의 앞의 5개 값을 확인합니다.
df_books[text_columns].head()

,상품명,저자,출판사,분야
0,소년이 온다,한강,창비,소설
1,모순,양귀자,쓰다,소설
2,결국 국민이 합니다,이재명,오마이북,정치/사회
3,혼모노,성해나,창비,소설
4,급류,정대건,민음사,소설


### 🧹 문자열 공백 제거 결과 요약

* **처리 완료**: 지정한 4개 컬럼(`상품명`, `저자`, `출판사`, `분야`) 데이터의 양끝에 숨어있던 불필요한 공백(띄어쓰기)이 모두 성공적으로 제거되었습니다.
* **데이터 확인**: 상위 5개 행을 확인해 본 결과, 텍스트가 밀리거나 이상한 여백 없이 깔끔하게 정돈된 것을 볼 수 있습니다.
* **의미**: 겉보기엔 큰 차이가 없지만, 데이터 분석에서는 아주 중요한 작업입니다! 이제 컴퓨터가 단어를 정확하게 인식하게 되어, 앞으로 '소설' 분야의 책이 몇 권인지 세거나 특정 저자의 책을 검색할 때 띄어쓰기 때문에 누락되는 오류를 방지할 수 있습니다.

### 실습 9. 중복 데이터 확인하기

 

데이터에서 중복을 발견했다고 해서 바로 삭제하면 안 됩니다.

 

먼저 무엇을 기준으로 중복인지 생각해야 합니다.

 

예를 들어 같은 상품명이 여러 번 등장할 수 있습니다.

 

하지만 같은 제목이라도 다른 판본, 다른 상품, 개정판 등이 있을 수 있습니다.

 

따라서 이번 데이터에서는 상품명만 보고 중복 행을 바로 삭제하지 않습니다.

 

우선 상품을 구분하는 판매상품ID를 확인합니다.

In [22]:
# 1. 전체 행(모든 컬럼의 값)이 완전히 똑같은 중복 데이터 개수 확인
total_duplicates = df_books.duplicated().sum()
print(f"전체 행 중복 개수: {total_duplicates}개")

# 2. '판매상품ID'만 기준으로 중복된 데이터 개수 확인
id_duplicates = df_books.duplicated(subset=['판매상품ID']).sum()
print(f"판매상품ID 중복 개수: {id_duplicates}개")

# 3. '상품명'만 기준으로 중복된 데이터 개수 확인
name_duplicates = df_books.duplicated(subset=['상품명']).sum()
print(f"상품명 중복 개수: {name_duplicates}개")

전체 행 중복 개수: 0개
판매상품ID 중복 개수: 0개
상품명 중복 개수: 0개


### 실습 10. 전처리 결과 검증하기

지금까지 여러 전처리를 수행했습니다.

마지막에는 반드시 데이터가 의도한 형태가 되었는지 다시 확인합니다.

In [23]:
print("=== 1. 데이터 크기 (shape) ===")
print(df_books.shape)

print("\n=== 2. 컬럼 목록 (columns) ===")
print(df_books.columns)

print("\n=== 3. 컬럼별 자료형 (dtypes) ===")
# 판매가가 int(정수형)인지, 발행일이 datetime(날짜형)인지 확인합니다.
print(df_books.dtypes)

print("\n=== 4. 결측치 개수 (isna().sum()) ===")
# 빈칸(결측치)이 모두 0개인지 확인합니다.
print(df_books.isna().sum())

print("\n=== 5. 판매가 컬럼 (앞의 5개) ===")
# 쉼표(,)가 사라졌는지 확인합니다.
print(df_books['판매가'].head())

print("\n=== 6. 발행일 컬럼 (앞의 5개) ===")
# 2026-07-01 처럼 날짜 형태로 잘 나오는지 확인합니다.
print(df_books['발행일'].head())

print("\n=== 7. 판매상품ID 중복 개수 ===")
print(df_books.duplicated(subset=['판매상품ID']).sum())

print("\n=== 8. 전체 데이터 확인 (앞의 5행) ===")
# 주피터 노트북에서 표 형태로 예쁘게 보기 위해 맨 마지막 줄에 적습니다.
df_books.head()

=== 1. 데이터 크기 (shape) ===
(199, 8)

=== 2. 컬럼 목록 (columns) ===
Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

=== 3. 컬럼별 자료형 (dtypes) ===
순위                 int64
판매상품ID               str
상품명                  str
판매가                int64
저자                   str
출판사                  str
발행일       datetime64[us]
분야                   str
dtype: object

=== 4. 결측치 개수 (isna().sum()) ===
순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

=== 5. 판매가 컬럼 (앞의 5개) ===
0    13500
1    11700
2    19800
3    16200
4    12600
Name: 판매가, dtype: int64

=== 6. 발행일 컬럼 (앞의 5개) ===
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]

=== 7. 판매상품ID 중복 개수 ===
0

=== 8. 전체 데이터 확인 (앞의 5행) ===


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,2014-05-19,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,2013-04-01,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,2025-04-15,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,2025-03-28,소설
4,5,S000200550189,급류,12600,정대건,민음사,2022-12-22,소설


### 🎯 최종 데이터 전처리 완료 요약

지금까지 진행한 모든 데이터 정제 작업이 완벽하게 적용되었습니다!

* **데이터 크기 및 상태**: 총 199개의 도서 데이터가 8개의 핵심 컬럼으로 깔끔하게 정리되었습니다.
* **결측치 및 중복 확인**: 비어있는 값(결측치)도 없고, 겹치는 데이터(중복)도 없는 **아주 깨끗한 상태**입니다.
* **자료형 변환 성공**:
  * `판매가`: 쉼표가 제거되고 계산 가능한 **숫자(int64)**로 잘 바뀌었습니다.
  * `발행일`: 컴퓨터가 인식할 수 있는 **날짜(datetime64)** 형태(YYYY-MM-DD)로 완벽하게 변환되었습니다.
* **의미**: 
  * 데이터를 분석하기 위한 귀찮고 복잡한 '준비 운동'이 아주 성공적으로 끝났습니다. 
  * 이제 이 완성된 데이터(`df_books`)를 활용해 분야별 도서 비율을 시각화하거나, 평균 가격을 계산하는 등 본격적이고 재밌는 데이터 분석을 바로 시작하실 수 있습니다! 🚀

### 실습 11. 전처리 데이터 저장하기

 

다음 Chapter에서는 이 데이터를 다시 사용합니다.

 

따라서 전처리한 DataFrame을 CSV 파일로 저장합니다.

 

추천 파일명은 다음과 같습니다.

In [24]:
# df_books 데이터를 CSV 파일로 저장하기
df_books.to_csv('book_bestseller_clean.csv', encoding='utf-8-sig', index=False)

### 📝 베스트셀러 데이터 전처리 완료 요약

**1. 무엇을 확인했나요?**
* **원본 데이터 크기**: 총 199행, 13열 `(199, 13)` 형태의 데이터였습니다.
* **결측치(빈칸)**: 확인 결과 '저자'와 '분야' 컬럼 모두 빈칸(결측치)이 0개였습니다.
* **중복 데이터**: '판매상품ID'를 기준으로 확인해 보니 중복된 행 역시 0개였습니다.

**2. 어떤 전처리를 했나요?**
* **컬럼 추출**: 전체 데이터 중 분석에 필요한 8개의 핵심 컬럼만 선택했습니다.
* **자료형 변환**:
  * `판매가`: 문자열에 포함된 쉼표(,)를 지우고 계산이 가능한 숫자형(`int64`)으로 변환했습니다.
  * `발행일`: 단순한 문자열/숫자를 진짜 날짜형(`datetime64[us]`)으로 변환했습니다.
* **문자열 정리**: 글자 양끝에 눈에 보이지 않게 숨어있던 공백(띄어쓰기)을 모두 지워 깔끔하게 정리했습니다.

**3. 다음 분석을 위해 어떤 상태가 되었나요?**
* 비어있거나 겹치는 데이터가 없고, 가격은 숫자로, 날짜는 날짜로 알맞게 세팅되었습니다. 
* 이제 오류 걱정 없이 분야별 통계를 내거나 가격을 비교하는 등 본격적인 데이터 분석을 바로 시작할 수 있는 준비된 상태입니다!